db/schema.sql

```diff
@@ -0,0 +1,92 @@
+    task_id            BYTEA NOT NULL,                  -- task
+    aggregation_param  BYTEA NOT NULL,                  -- encoded aggregation parameter (opaque VDAF message)
+    state              AGGREGATION_JOB_STATE NOT NULL,  -- current state of the aggregation job
+
+    CONSTRAINT fk_task_id FOREIGN KEY(task_id) REFERENCES tasks(id)
+);
+
+-- An aggregation attempt for a single client report. An aggregation job logically contains a number
+-- of report aggregations. A single client report might be aggregated in multiple aggregation jobs &
+-- therefore have multiple associated report aggregations.
+CREATE TABLE report_aggregations(
+    id                  BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY, -- artificial ID, internal-only
+    aggregation_job_id  BIGINT NOT NULL,  -- the aggregation job ID this report aggregation is associated with
+    client_report_id    BIGINT NOT NULL,  -- the client report ID this report aggregation is associated with
+    ord                 BIGINT NOT NULL,  -- a value used to specify the ordering of client reports in the aggregation job
+    transition          BYTEA NOT NULL,   -- encoded Transition message, representing the current preparation state of this report in this aggregation job
+
+    CONSTRAINT unique_ord UNIQUE(aggregation_job_id, ord),
+    CONSTRAINT fk_aggregation_job_id FOREIGN KEY(aggregation_job_id) REFERENCES aggregation_jobs(id),
+    CONSTRAINT fk_client_report_id FOREIGN KEY(client_report_id) REFERENCES client_reports(id)
+);
+CREATE INDEX report_aggregations_aggregation_job_id_index ON report_aggregations(aggregation_job_id);
+
+-- Information on incremental aggregation, for VDAFs that support incremental aggregation (eg prio3).
+CREATE TABLE batch_window_aggregations(
+    task_id             BYTEA NOT NULL,      -- the task ID
+    batch_window_start  TIMESTAMP NOT NULL,  -- the start of the batch window
+    aggregate_share     BYTEA NOT NULL,      -- the (incremental) aggregate share
+    report_count        BIGINT NOT NULL,     -- the (incremental) client report count
+    checksum            BYTEA NOT NULL,      -- the (incremental) checksum
+
+    PRIMARY KEY(task_id, batch_window_start)
+
+    -- TODO(brandon): decide how to count collection attempts: in this table? via collect_jobs inspection? somehow else?
+);
+
+CREATE TABLE collect_jobs(
+    id                    UUID DEFAULT gen_random_uuid() PRIMARY KEY, -- UUID used by collector to refer to this job
+    task_id               BYTEA NOT NULL,      -- the task ID being collected
+    batch_interval_start  TIMESTAMP NOT NULL,  -- the start of the batch interval
+    batch_interval_end    TIMESTAMP NOT NULL,  -- the end of the batch interval
+    aggregation_param     BYTEA NOT NULL,      -- the aggregation parameter (opaque VDAF message)
+
+    CONSTRAINT fk_task_id FOREIGN KEY(task_id) REFERENCES tasks(id)
+);
```

Comment:
Presumably, this is to check if there is a `collect_job` covering the report's batch?

I left a TODO for this -- I slightly prefer adding a `collect_count` column to the `batch_aggregations` table. But this preference comes from wanting to use the DB as close to a KV store as possible (i.e. looking things up by direct key, rather than a range), with the hope that simplicity of use will result in consistent (good) performance. And using an index on `collect_jobs` is definitely the more SQL-y way to do things and offloads more of the work onto the SQL engine. And we'll already need range lookups for other things (e.g. finding client reports to add to an aggregation job).

I'd be interested if Tim has thoughts; if not, I suppose I'll take your vote for an index as a strong enough indication that we should go that way.

Reply:
At the risk of being unhelpful: I have little to no intuition or insight about database schemas or common storage performance pitfalls, I'm afraid, so I vote for doing the simplest thing we possibly can that will work and iterating on it as we learn more.

Reply:
OK, the simplest thing will be an index (in terms of logic we have to write/maintain), so I'm going with that for now.

Reply:
Yes, that's what I was thinking. The per-epoch counters would work, in that case we would have to increment one or more counters with the creation of a collect_jobs row, and during upload we would check if the counter column is zero, or if there is no corresponding row for the task and epoch.

I think it's safe to assume the dominant action in our aggregator will be report upload. Here, the two options we're looking at mean the "late upload check" reads will either contend with writes to either `collect_jobs` or `batch_window_aggregations`. We will have significantly more writes to `batch_window_aggregations`, to update the three incremental aggregation columns, but not the `collect_count` we're considering. I don't know how the database's concurrency control would handle UPDATEs on one set of cell vs. reads on a disjoint cell in the same row. Based on this intuition, I think indexing `collect_jobs` and querying that would be the best way to do the "late upload check", but I'm fine with picking either to iterate and experiment.

(As discussed in another thread, we may reconsider how `batch_window_aggregations` works in the future. In particular, I note that the current design will serialize our incremental aggregations, so if two aggregator processes contend on updating the same window, they could wind up rolling back and having to re-do field element math)